src/librustc_mir/interpret/place.rs

```diff
@@ -137,7 +137,7 @@ impl<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M> {
             Value::Scalar(_) if offset.bytes() == 0 && field.size == base_layout.size => Ok(Some((base, field.ty))),
             // split fat pointers, 2 element tuples, ...
             Value::ScalarPair(a, b) if base_layout.fields.count() == 2 => {
-                let val = [a, b][field_index];
+                let val = [a, b][base_layout.fields.memory_index(field_index)];
```

Comment:
This is still not a good idea (if you want to see what happens, try adding a bunch of ZSTs) - what you want is to check whether `offset.bytes() == 0` or not. I could've sworn I already wrote the code in this module to do it correctly, are there several similar pieces of code in here?